Analytics/Integrations/Kahuna/SEGKahunaIntegration.m

```diff
@@ -38,42 +103,400 @@ - (void)validate {
 #pragma mark - Analytics API
 
 - (void)identify:(NSString *)userId traits:(NSDictionary *)traits options:(NSDictionary *)options {
-  [KahunaAnalytics setUserCredentialsWithKey:KAHUNA_CREDENTIAL_USERNAME andValue:userId];
-  NSString *email = [self.class extractEmail:userId traits:traits];
-  if (email) {
-    [KahunaAnalytics setUserCredentialsWithKey:KAHUNA_CREDENTIAL_EMAIL andValue:email];
+  @try {
+    NSMutableDictionary *attributes = [[NSMutableDictionary alloc] init];
+    if (KAHUNA_NOT_STRING_NULL_EMPTY (userId)) {
+      [KahunaAnalytics setUserCredentialsWithKey:KAHUNA_CREDENTIAL_USER_ID andValue:userId];
+    }
+    
+    // We will go through each of the above keys, and try to see if the traits has that key. If it does, then we will add the key:value as a credential.
+    // All other traits is being tracked as an attribute.
+    for (NSString *eachKey in traits) {
+      if (!KAHUNA_NOT_STRING_NULL_EMPTY (eachKey)) continue;
+      
+      NSString *eachValue = [traits objectForKey:eachKey];
+      if (KAHUNA_NOT_STRING_NULL_EMPTY (eachValue)) {
+        // Check if this is a Kahuna credential key.
+        if ([_kahunaCredentialsKeys containsObject:eachKey]) {
+          [KahunaAnalytics setUserCredentialsWithKey:eachKey andValue:eachValue];
+        } else {
+          [attributes setValue:eachValue forKey:eachKey];
+        }
+      } else if ([eachValue isKindOfClass:[NSNumber class]]) {
+        // Check if this is a Kahuna credential key.
+        if ([_kahunaCredentialsKeys containsObject:eachKey]) {
+          [KahunaAnalytics setUserCredentialsWithKey:eachKey andValue:[NSString stringWithFormat:@"%@", eachValue]];
+        } else {
+          [attributes setValue:[NSString stringWithFormat:@"%@", eachValue] forKey:eachKey];
+        }
+      } else {
+        @try {
+          [attributes setValue:[eachValue description] forKey:eachKey];
+        }
+        @catch (NSException *exception) {
+          // Do nothing.
+        }
+      }
+    }
+    
+    // Track the attributes if we have any items in it.
+    if (attributes.count > 0) {
+      [KahunaAnalytics setUserAttributes:attributes];
+    }
+  }
+  @catch (NSException *exception) {
```

Comment:
Defensive in order to ensure all the key:values of the traits dictionary do not change to un-anticipated types.